src/mgr/DaemonServer.cc

```diff
@@ -371,64 +373,107 @@ bool DaemonServer::_allowed_command(
   return capable;
 }
 
-class ReplyOnFinish : public Context {
-  DaemonServer* mgr;
-  MCommand *m;
-  bufferlist odata;
-
-public:
-  bufferlist from_mon;
-  string outs;
-
-  ReplyOnFinish(DaemonServer* mgr, MCommand *m, bufferlist&& odata)
-    : mgr(mgr), m(m), odata(std::move(odata))
-  {}
-  void finish(int r) override {
-    odata.claim_append(from_mon);
-    mgr->_reply(m, r, outs, odata);
-  }
-};
-
 bool DaemonServer::handle_command(MCommand *m)
 {
   int r = 0;
   std::stringstream ss;
-  bufferlist odata;
   std::string prefix;
 
   assert(lock.is_locked_by_me());
 
-  cmdmap_t cmdmap;
+  /**
+   * The working data for processing an MCommand.  This lives in
+   * a class to enable passing it into other threads for processing
+   * outside of the thread/locks that called handle_command.
+   */
+  class CommandContext
+  {
+    public:
+    MCommand *m;
+    ConnectionRef con;
+    bufferlist odata;
+    cmdmap_t cmdmap;
+
+    CommandContext(MCommand *m_)
+      : m(m_)
+    {
+      con = m->get_connection();
```

Comment:
no need to store the `con` as a member variable, will elaborate the rationale below.